### ISSUE TITLE
Prom2: Protobuf Exposition Format Support?

### ISSUE THREAD

**Issue by snarlysodboxer** [prometheus/prometheus#2788]
There seems to be [some question](https://github.com/prometheus/client_model/issues/13#issuecomment-304614153) as to whether the Protobuf Exposition Format will be supported in Prometheus 2.0. If it's not going to be supported, we assume we should stop developing Prometheus 1.X Protobuf code in all client libraries.

Thanks!

**Comment by fabxc** [comment]
The current tendency (but no final call) is that text format works really well and allows us to do some fancy ingestion optimisations.
Currently Prometheus 2.0 cannot ingest protobuf anymore. Do you have a an application or exporter that can only export the protobuf format?

@beorn7 @brian-brazil @juliusv

**Comment by brian-brazil** [comment]
I am aware of one (internal) application that's proto only, but once client_golang supports timestamps I presume it'll switch over.

**Comment by beorn7** [comment]
I have a number of concerns about dropping protobuf support, beyond the question if client libraries support the text format or not. I'm not sure if a GH issue is the right place to make the final call. Perhaps open a shared Google doc and develop that into a design-doc/RFC? @fabxc It would be great if you could document the ingestion optimizations there.

**Comment by kongchen** [comment]
Some of our service's metrics size reaches 10MB in text mode and we're currently moving to pb, cannot imaging pb is not supported in future.

**Comment by fabxc** [comment]
@kongchen from what we have gathered so far, the wire size does not substantially differ between protobuf and text format+gzip.

**Comment by beorn7** [comment]
In fact, gzip'd text format is smaller than uncompressed proto format. However, usually both are compressed, and then the size is really not much different.

Things changes in scenarios where compression is not an option. With the dramatically increased ingestion efficiency of Prom2, I can imagine that the cost of compression becomes relevant even on the side of the Prometheus server. (On the side of the monitored target, it's even more likely that compression is a relevant cost.)

**Comment by kongchen** [comment]
@fabxc @beorn7 Thanks, we can use gzip for compression by enable server side gzip(tomcat actually). 

However, now the only concern is that will this lead Prometheus server to a high cpu usage? After all we have thousands of nodes for the server to pull.

**Comment by beorn7** [comment]
Prometheus 2 uses dramatically less CPU for ingestion anyway. The CPU you need for compression is very small compared to that.

(What I meant above is that the CPU cycles needed for compression only become relevant _because_ Prom2 uses so little CPU for the rest of the ingestion.)

**Comment by kongchen** [comment]
Thanks, look forward to it!

**Comment by SimenB** [comment]
Speaking of prometheus v2, is there an issue/document with what it'll contain/change? I see the milestone (https://github.com/prometheus/prometheus/milestone/4) but no general issue 🙂

**Comment by juliusv** [comment]
@SimenB Let's keep this issue on topic (please ask general questions on the mailing lists or chat), but https://prometheus.io/blog/2017/06/21/prometheus-20-alpha3-new-rule-format/ covers some of that.

**Comment by SimenB** [comment]
Now that Prometheus 2 is in RC, should we conclude that protobuf support is out?

**Comment by beorn7** [comment]
Not sure if there was a final call, but at another occasion, I wrote something down about text vs. protobuf: https://github.com/RichiH/OpenMetrics/blob/master/protobuf_vs_text.md

**Comment by zegelin** [comment]
Can the documentation be updated to reflect the decision to remove protobuf support?

Currently https://prometheus.io/docs/instrumenting/exposition_formats/ lists protobuf as "Supported in | Prometheus version >=0.4.0", yet version 2.1.0 doesn't support protobuf.

A custom exporter application I wrote used the protobuf format. After I upgraded to 2.1.0 I was wondering why I was getting errors `err="no token found"` in the logs. I had to do a bit of digging before I realised the documentation is wrong.
